tests/test_preprocessing.py

```diff
@@ -0,0 +1,169 @@
+
+    def test_clean_html_and_symbols(self):
+        input_text = "<p>This is a test.</p> \u1234 \uabcd"
+        expected_output = "This is a test.    "
+        result = clean_html_and_symbols(input_text)
+        self.assertEqual(result, expected_output)
+
+    def test_extract_code_snippets(self):
+        input_text = (
+            "Here is some code:\n"
+            "```python\n"
+            "def hello():\n"
+            "    print('Hello World')\n"
+            "```\n"
+            "End of code."
+        )
+        expected_cleaned_text = "Here is some code:\n\nEnd of code."
+        expected_code_snippets = ["def hello():\n    print('Hello World')"]
+        cleaned_text, code_snippets = extract_code_snippets(input_text)
+        self.assertEqual(cleaned_text, expected_cleaned_text)
+        self.assertEqual(code_snippets, expected_code_snippets)
+
+    def test_extract_images(self):
+        input_text = "This is an image: ![Alt Text](http://example.com/image.png)"
+        expected_cleaned_text = "This is an image: "
+        expected_images = [("Alt Text", "http://example.com/image.png")]
+        images, cleaned_text = extract_images(input_text)
+        self.assertEqual(cleaned_text, expected_cleaned_text)
+        self.assertEqual(images, expected_images)
+
+    def test_extract_links(self):
+        input_text = "This is a link: [Example](http://example.com)"
+        expected_cleaned_text = "This is a link: Example"
+        expected_links = [("Example", "http://example.com")]
+        links, cleaned_text = extract_links(input_text)
+        self.assertEqual(cleaned_text, expected_cleaned_text)
+        self.assertEqual(links, expected_links)
+
+    def test_extract_tables(self):
+        input_text = (
+            "Here is a table:\n\n"
+            "| Header1 | Header2 |\n"
+            "|---------|---------|\n"
+            "| Cell1   | Cell2   |\n"
+            "| Cell3   | Cell4   |\n"
+            "End of table."
+        )
+        expected_cleaned_text = "Here is a table:\n\nEnd of table."
+        expected_tables = [
+            [
+                ["Header1", "Header2"],
+                ["Cell1", "Cell2"],
+                ["Cell3", "Cell4"]
+            ]
+        ]
+        tables, cleaned_text = extract_tables(input_text)
+        self.assertEqual(cleaned_text, expected_cleaned_text)
+        self.assertEqual(tables, expected_tables)
+
+    def test_extract_markdown_elements(self):
+        input_text = (
+            "Sample text with markdown elements.\n\n"
+            "![Image Alt](http://example.com/image.png)\n"
+            "[Link Text](http://example.com)\n"
+            "```java\n"
+            "public class HelloWorld {\n"
+            "    public static void main(String[] args) {\n"
+            "        System.out.println(\"Hello, World\");\n"
+            "    }\n"
+            "}\n"
+            "```\n"
+            "| Header1 | Header2 |\n"
+            "|---------|---------|\n"
```

Comment:
Maybe add the test case where the tables are aligned left, right and center.

The tables are aligned
- left: when the line is `|-------|` or `|:------|`
- center: `|:--------:|`
- right: `|--------:|`